src/Microsoft.TestPlatform.CrossPlatEngine/PostProcessing/ArtifactProcessingManager.cs

```diff
@@ -0,0 +1,219 @@
+        {
+            EqtTrace.Verbose("ArtifactProcessingManager.PostProcessArtifacts: There are no artifacts to postprocess");
+            return;
+        }
+
+        var testArtifacts = LoadTestArtifacts();
+        if (testArtifacts?.Length > 0)
+        {
+            try
+            {
+                await MergeDataCollectorAttachments(testArtifacts);
+            }
+            finally
+            {
+                _fileHelper.DeleteDirectory(_processArtifactFolder, true);
+                EqtTrace.Verbose($"ArtifactProcessingManager.PostProcessArtifacts: Directory '{_processArtifactFolder}' removed.");
+            }
+        }
+        else
+        {
+            EqtTrace.Warning($"ArtifactProcessingManager.PostProcessArtifacts: There are no artifacts to postprocess also if the artifact directory '{_processArtifactFolder}' exits");
+        }
+    }
+
+
+    private async Task MergeDataCollectorAttachments(TestArtifacts[] testArtifacts)
+    {
+        // We take the biggest runsettings in size, it should be the one with more configuration.
+        // In future we can think to merge...but it's not easy for custom config, we could break something.
+        string runsettingsFile = testArtifacts
+            .SelectMany(x => x.Artifacts.Where(x => x.Type == ArtifactType.Runsettings))
+            .Select(x => new FileInfo(x.FileName))
+            .OrderByDescending(x => x.Length)
+            .FirstOrDefault()?
+            .FullName;
+
+        string runsettingsXml = null;
+        if (runsettingsFile is not null)
+        {
+            using var artifactStream = _fileHelper.GetStream(runsettingsFile, FileMode.Open, FileAccess.Read);
+            using var streamReader = new StreamReader(artifactStream);
+            runsettingsXml = await streamReader.ReadToEndAsync();
+            EqtTrace.Verbose($"ArtifactProcessingManager.MergeDataCollectorAttachments: Chosen runsettings \n{runsettingsXml}");
+        }
+        else
+        {
+            EqtTrace.Verbose($"ArtifactProcessingManager.MergeDataCollectorAttachments: Null runsettings");
+        }
+
+        HashSet<InvokedDataCollector> invokedDataCollectors = new();
+        List<AttachmentSet> attachments = new();
+        foreach (var artifact in testArtifacts
+            .SelectMany(x => x.Artifacts)
+            .Where(x => x.Type == ArtifactType.ExecutionComplete))
+        {
+            using var artifactStream = _fileHelper.GetStream(artifact.FileName, FileMode.Open, FileAccess.Read);
+            using var streamReader = new StreamReader(artifactStream);
+            string executionCompleteMessage = await streamReader.ReadToEndAsync();
+            EqtTrace.Verbose($"ArtifactProcessingManager.MergeDataCollectorAttachments: ExecutionComplete message \n{executionCompleteMessage}");
+            TestRunCompleteEventArgs eventArgs = _dataSerialized.DeserializePayload<TestRunCompleteEventArgs>(_dataSerialized.DeserializeMessage(executionCompleteMessage));
```

Comment:
add a comment to avoid try/catch for loading...we want to have an error instead of partial result, users could be confused...we're ok but with wrong result.